mesonbuild/modules/fs.py

```diff
@@ -189,39 +189,64 @@ def is_samepath(self, state: 'ModuleState', args: T.Tuple['FileOrString', 'FileO
+    @typed_pos_args('fs.parent', (str, File, CustomTarget, CustomTargetIndex, BuildTarget))
+    def parent(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
+        arg = args[0]
+        if isinstance(arg, str):
+            original = PurePath(arg)
+        elif isinstance(arg, File):
+            FeatureNew('fs.parent(file)', '0.59.0').use(state.subproject, location=state.current_node)
+            original = PurePath(str(arg))
+        else:
+            FeatureNew('fs.parent(build_tgt | custom_tgt | custom_idx)', '1.3.0').use(state.subproject, location=state.current_node)
+            original = PurePath(state.backend.get_target_filename(arg))
         new = original.parent
         return str(new)
 
     @noKwargs
-    @typed_pos_args('fs.name', (str, File))
-    def name(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: T.Dict[str, T.Any]) -> str:
-        if isinstance(args[0], File):
-            FeatureNew('fs.name_file', '0.59.0').use(state.subproject)
-        original = PurePath(str(args[0]))
+    @typed_pos_args('fs.name', (str, File, CustomTarget, CustomTargetIndex, BuildTarget))
+    def name(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
+        arg = args[0]
+        if isinstance(arg, str):
+            original = PurePath(arg)
+        elif isinstance(arg, File):
+            FeatureNew('fs.name(file)', '0.59.0').use(state.subproject, location=state.current_node)
+            original = PurePath(str(arg))
+        else:
+            FeatureNew('fs.name(build_tgt | custom_tgt | custom_idx)', '1.3.0').use(state.subproject, location=state.current_node)
+            original = PurePath(state.backend.get_target_filename(arg))
         new = original.name
         return str(new)
 
     @noKwargs
-    @typed_pos_args('fs.stem', (str, File))
+    @typed_pos_args('fs.stem', (str, File, CustomTarget, CustomTargetIndex, BuildTarget))
     @FeatureNew('fs.stem', '0.54.0')
-    def stem(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: T.Dict[str, T.Any]) -> str:
-        if isinstance(args[0], File):
-            FeatureNew('fs.stem_file', '0.59.0').use(state.subproject)
-        original = PurePath(str(args[0]))
+    def stem(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
+        arg = args[0]
+        if isinstance(arg, str):
+            original = PurePath(arg)
+        elif isinstance(arg, File):
+            FeatureNew('fs.stem(file)', '0.59.0').use(state.subproject, location=state.current_node)
+            original = PurePath(str(arg))
+        else:
+            FeatureNew('fs.stem(build_tgt | custom_tgt | custom_idx)', '1.3.0').use(state.subproject, location=state.current_node)
+            original = PurePath(state.backend.get_target_filename(arg))
```

Comment:
This seems all quite redundant - perhaps worth creating new types to deduplicate typed arguments, and moving the type specific handling into a common function that gets the function name as an argument?